dartagnan/src/main/java/com/dat3m/dartagnan/verification/solving/RefinementSolver.java

```diff
@@ -537,14 +543,107 @@ private static void addBiases(Wmm memoryModel, EnumSet<Baseline> biases) {
+
+            @Override
+            public Void visitAxiom(Axiom axiom) {
+                polarity = !axiom.isNegated();
+                process(axiom);
+                return axiom.getRelation().getDefinition().accept(this);
+            }
+
+            @Override
+            public Void visitDefinition(Definition def) {
+                if (process(def)) {
+                    def.getConstrainedRelations().subList(1, def.getConstrainedRelations().size())
+                            .forEach(r -> r.getDefinition().accept(this));
+                }
+                return null;
+            }
+
+            @Override
+            public Void visitDifference(Difference def) {
+                if (process(def)) {
+                    def.getMinuend().getDefinition().accept(this);
+                    polarity = !polarity;
+                    def.getSubtrahend().getDefinition().accept(this);
+                    polarity = !polarity;
+                }
+                return null;
+            }
+
+            private boolean process(Constraint c) {
+                return (polarity ? positives : negatives).add(c);
+            }
+        };
+
+        wmm.getAxioms().forEach(ax -> ax.accept(collector));
+        return new PolaritySeparator(positives, negatives);
+    }
+
+    /*
+        This pass rewrites the memory model so that no positive constraints appear on the rhs-side of a difference.
+        Suppose "r = a \ b" where r and a are negative and b is positive.
+        For example, this is the case if there is a constraint "~empty(r)"
+        We rewrite the model to
+            let b`= free()
+            let r = a \ b' // b' gets cut, but it is effectively a base relation so this causes no large encoding.
+            empty (r & b) // b is now in a positive position and thus not need to get cut
+     */
+    private void instrumentPolaritySeparation(Wmm wmm) {
+        final PolaritySeparator separator = computePolaritySeparator(wmm);
+        final Set<Difference> negDiff = separator.negatives().stream()
+                .filter(Difference.class::isInstance).map(Difference.class::cast)
+                .filter(diff -> !separator.negatives().contains(diff.getSubtrahend().getDefinition()))
+                .collect(Collectors.toSet());
+
+        final Map<Relation, Relation> replacements = new HashMap<>();
+        int counter = 0;
+        for (Difference diff : negDiff) {
+            // r = a \ b where r and a are negative and b is positive.
+            final Relation r = diff.getDefinedRelation();
+            final Relation a = diff.getMinuend();
+            final Relation b = diff.getSubtrahend();
+
+            // (1) Create b' (if not already existing)
+            final Relation bPrime;
+            if (replacements.containsKey(b)) {
+                bPrime = replacements.get(b);
+            } else {
+                final String bPrimeName = b.getName().map(n -> n + "#POS").orElse("__POS" + counter++);
```

Comment:
Why the different `#POS` and `__POS`?

Reply:
Oh, a mistake.

Reply:
Actually, this was intended. I like to put the prefix `__` for all internally generated stuff and use `#` for annotations/decorations of existing stuff.

Reply:
Is it the case that if the relation was named, you add `#POS` and if the relation was unnamed you add `__POS`?

Reply:
Yes. In the latter case I also add a number to avoid duplicated names.